pandas/core/algorithms.py

```diff
@@ -2055,13 +2056,52 @@ def sort_mixed(values):
         strs = np.sort(values[str_pos])
         return np.concatenate([nums, np.asarray(strs, dtype=object)])
 
+    def sort_tuples(values):
+        # sorts tuples with mixed values. can handle nan vs string comparisons.
+        def cmp_func(index_x, index_y):
+            x = values[index_x]
+            y = values[index_y]
+            if x == y:
+                return 0
+            len_x = len(x)
+            len_y = len(y)
+            for i in range(max(len_x, len_y)):
```

Comment:
generally speaking I wonder if the control flow can be simplified a bit here. you have a number of ifs but only two possible outputs

Reply:
3 outcomes (=,<,>), but in combination with nan, they increase.

there are 9 ifs of which 2 are not relevant (shortcut before the loop, fall-though after the loop).

7 ifs remaining:

1. equality case (continue)
2. greater than case
3. smaller than case
4. x is nan, but y is not 
5. y is nan, but x is not
6. xvec is smaller than yvec
7. yvec is smaller than xvec

I could merge some of the cases (with same return value), but that would compromise readability and require nan checks to be done earlier.